src/pip/_internal/metadata/base.py

```diff
@@ -73,6 +76,28 @@ def location(self) -> Optional[str]:
         """
         raise NotImplementedError()
 
+    @property
+    def editable_project_location(self) -> Optional[str]:
+        """The project location for editable distributions.
+
+        This is the directory where pyproject.toml or setup.py is located.
+        None if the distribution is not installed in editable mode.
+        """
+        # TODO: this property is relatively costly to compute, memoize it ?
+        direct_url = self.direct_url
+        if direct_url:
+            if isinstance(direct_url.info, DirInfo) and direct_url.info.editable:
```

Comment:
There is a name property so we could do `if direct_url.info.name == DirInfo.name and direct_url.info.editable`, but I'm not sure it's more readable. And isinstance is the pattern we use elswhere we need to dispatch based on the direct_url.info type.

That is be a case where 3.10 structural pattern matching would help, but for now I'd leave it like this for consistency ?

Reply:
The `isinstance` version is more readable, but `direct_url.info.name == DirInfo.name` does not really solve the structural issue. The problem is that both methods require the call site to have knowledge about `DirInfo` being a supertype/protocol/union, breaking the type abstraction. If we add an additional direct URL type in the future that’s also local but not a directory, we’d need to also fix the type check here, which means the abstraction isn’t really fulfilling its purpose. An “proper” OOP design would be something like

```python
if direct_url.is_local() and direct_url.info.editable
```

or even

```python
if direct_url.is_local_editable()
```

so a new direct URL type can implement without checking the call sites.

But then againm it’s quite improbable we’d ever add the hypothetical direct URl type, so I’m not very strong about this. It’s more like a coding style thing than actually beneficial.

Reply:
I'm not sure I buy the OOP design argument because there is nothing common between the 3 `info` types so nothing to abstract. Factoring out `is_local_editable` makes sense, though, so I did that.